src/service/gateways/bitstamp.ts

```diff
@@ -0,0 +1,360 @@
+    
+    public postToEndpoint = <TRequest, TResponse>(endpoint: string, req : TRequest) : Q.Promise<TResponse> => {
+        return this.requestFromEndpoint<TResponse>({
+            url: this.makeUrl(endpoint),
+            form: this.addAuthentication(req),
+            method: 'POST',
+            timeout: 5000,
+            headers: { 'User-Agent': BitstampAuthenticatedClient.UserAgent }
+        });
+    }; 
+    
+    public getFromEndpoint = <TResponse>(endpoint: string) => {
+        return this.requestFromEndpoint<TResponse>({
+            url: this.makeUrl(endpoint),
+            method: 'GET',
+            timeout: 5000,
+            headers: { 'User-Agent': BitstampAuthenticatedClient.UserAgent }
+        });
+    };
+    
+    private requestFromEndpoint = <TResponse>(options: request.Options) => {
+        var defer = Q.defer<TResponse>();
+        request(options, (err, resp, body) => {
+            if (err) defer.reject(err);
+            else defer.resolve(JSON.parse(body));
+        });
+        return defer.promise;
+    };
+}
+
+class BitstampEntryGateway implements Interfaces.IOrderEntryGateway {
+    OrderUpdate = new Utils.Evt<Models.OrderStatusReport>();
+    ConnectChanged = new Utils.Evt<Models.ConnectivityStatus>();
+    
+    public cancelsByClientOrderId = false;
+
+    cancelOrder = (cancel : Models.BrokeredCancel) : Models.OrderGatewayActionReport => {
+        this._client.postToEndpoint<{}, boolean>("cancel_order/", {id: cancel.exchangeId}).then(ack => {
+            if (ack) this.OrderUpdate.trigger({
+                exchangeId: cancel.exchangeId, 
+                orderStatus: Models.OrderStatus.Cancelled
+            });
+            else this.OrderUpdate.trigger({
+                exchangeId: cancel.exchangeId, 
+                orderStatus: Models.OrderStatus.Rejected, 
+                cancelRejected: true
+            });
+        }).done();
+        return new Models.OrderGatewayActionReport(Utils.date());
+    };
+
+    replaceOrder = (replace : Models.BrokeredReplace) : Models.OrderGatewayActionReport => {
+        this.cancelOrder(new Models.BrokeredCancel(replace.origOrderId, replace.orderId, replace.side, replace.exchangeId));
+        return this.sendOrder(replace);
+    };
+
+    sendOrder = (order : Models.BrokeredOrder) : Models.OrderGatewayActionReport => {
+        var side = order.side === Models.Side.Bid ? "buy/" : "sell/";
+        this._client.postToEndpoint<{}, NewOrderAck>(side, {amount: order.quantity, price: order.price}).then(ack => {
+            this.OrderUpdate.trigger({
+                orderId: order.orderId, 
+                exchangeId: ack.id, 
+                orderStatus: Models.OrderStatus.Working
+            });
+        }).done();
+        return new Models.OrderGatewayActionReport(Utils.date());
+    };
+    
+    generateClientOrderId = () : string => shortId.generate();
+
+    private downloadOrderStatuses = () => {
+        // I really have no idea what to do here.
```

Comment:
I don't know how to get order status updates for live orders from the Bitstamp API. I assume you need to use `order_status` but the docs are not super clear about what they return. I'm interested here if any orders were executed (partially or completely) well after being submitted.